File: src/main/java/no/fintlabs/operator/FlaisApplicationSpec.java
package no.fintlabs.operator;

import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import io.fabric8.kubernetes.api.model.*;
import io.fabric8.kubernetes.api.model.apps.DeploymentStrategy;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import no.fintlabs.FlaisSpec;
import no.fintlabs.operator.kafka.Kafka;
import no.fintlabs.operator.onepassword.OnePassword;
import no.fintlabs.operator.pg.Database;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class FlaisApplicationSpec implements FlaisSpec {

    @JsonPropertyDescription("If present this application is consider to be a county application.")
    private String orgId;

    private List<EnvVar> env = new ArrayList<>();

    private List<EnvFromSource> envFrom = new ArrayList<>();

    private String image;
    private String imagePullPolicy = "IfNotPresent";

    private Integer replicas = 1;

    private ResourceRequirements resources = new ResourceRequirementsBuilder()
            .withRequests(new HashMap<>() {{
                put("cpu", new Quantity("250m"));
                put("memory", new Quantity("256"));
            }})
            .withLimits(new HashMap<>() {{
                put("cpu", new Quantity("500m"));
                put("memory", new Quantity("512Mi"));
            }})
            .build();

    private Integer port = 8080;

    private String restartPolicy = "Always";

    private DeploymentStrategy strategy;

    private Prometheus prometheus = new Prometheus();

    private OnePassword onePassword = new OnePassword();

    private Kafka kafka = new Kafka();

    private Database database = new Database();

}

File: src/main/kotlin/no/fintlabs/operator/api/v1alpha1/FlaisApplicationSpec.kt
package no.fintlabs.operator.api.v1alpha1

import io.fabric8.generator.annotation.Min
import io.fabric8.generator.annotation.Required
import io.fabric8.generator.annotation.ValidationRule
import io.fabric8.kubernetes.api.model.*
import io.fabric8.kubernetes.api.model.apps.DeploymentStrategy

data class FlaisApplicationSpec(
    @Required
    val orgId: String = "",

    @Min(0.0)
    val replicas: Int = 1,

    @Required
    val image: String = "",

    @ValidationRule(
    "self in ['IfNotPresent', 'Always', 'Never']",
    message = "Invalid imagePullPolicy, must be one of IfNotPresent, Always, Never"
    )
    val imagePullPolicy: String? = null,
    val imagePullSecrets: List<String> = emptyList(),

    val env: List<EnvVar> = emptyList(),
    val envFrom: List<EnvFromSource> = emptyList(),

    val resources: ResourceRequirements = ResourceRequirementsBuilder()
        .addToRequests("cpu", Quantity("250m"))
        .addToRequests("memory", Quantity("256Mi"))
        .addToLimits("cpu", Quantity("500m"))
        .addToLimits("memory", Quantity("512Mi"))
        .build(),

    @Min(1.0)
    val port: Int = 8080,

    @Deprecated("Does not exist on when applied on DeploymentSpec")
    @ValidationRule(
        "self in ['Always', 'OnFailure', 'Never']",
        message = "Invalid restartPolicy, must be one of Always, OnFailure, Never")
    val restartPolicy: String = "Always",

    val strategy: DeploymentStrategy? = null,

    @Deprecated("Use metrics in observability instead. Will be removed in future versions.")
    val prometheus: Metrics = Metrics(),
    val onePassword: OnePassword? = null,
    val kafka: Kafka = Kafka(),
    val database: Database = Database(),
    val url: Url = Url(),
    val ingress: Ingress? = null,
    val observability: Observability? = null
)